### PR TITLE
arb os upgrade for orbit chains

### PR DESCRIPTION
**Pull request by GreatSoshiant**

### CODE REVIEW

File: arbitrum-docs/launch-orbit-chain/concepts/arbos-upgrade.md
Comment (hkalodner) [inline arbitrum-docs/launch-orbit-chain/concepts/arbos-upgrade.md:2]: What do you think of re-forming this into a how to or potentially splitting it?

File: arbitrum-docs/launch-orbit-chain/concepts/arbos-upgrade.md
Comment (hkalodner) [inline arbitrum-docs/launch-orbit-chain/concepts/arbos-upgrade.md:28]: We should note that setting thing timestamp to 0 causes the upgrade to activate immediately

File: arbitrum-docs/launch-orbit-chain/concepts/arbos-upgrade.md
Comment (hkalodner) [inline arbitrum-docs/launch-orbit-chain/concepts/arbos-upgrade.md:32]: Nodes need to be fully backwards compatible with all arbos version the chain has run on so I think having the recommendation be to first issue the node release before anything else, but that people only need to actually make sure they've upgraded when the upgrade is activated.

Basically it's better to start by releasing the node upgrade to maximize how much time people have to upgrade

**Review by hkalodner — CHANGES_REQUESTED** [review]

File: arbitrum-docs/launch-orbit-chain/concepts/arbos-upgrade.md
Comment (DZGoldman) [inline arbitrum-docs/launch-orbit-chain/concepts/arbos-upgrade.md:26]: Maybe call this section "Schedule ArbOS _version_ Upgrade", with "ArbOS Upgrade" being the overarching term for the whole process, if that makes sense

File: arbitrum-docs/launch-orbit-chain/concepts/arbos-upgrade.md
Comment (Alirezaabbasi-hub) [reply arbitrum-docs/launch-orbit-chain/concepts/arbos-upgrade.md:2]: 👍

Comment (Alirezaabbasi-hub) [reply arbitrum-docs/launch-orbit-chain/concepts/arbos-upgrade.md:2]: > What do you think of re-forming this into a how to or potentially splitting it?

Comment (Alirezaabbasi-hub) [reply arbitrum-docs/launch-orbit-chain/concepts/arbos-upgrade.md:2]: 👍

Comment (symbolpunk) [reply arbitrum-docs/launch-orbit-chain/concepts/arbos-upgrade.md:2]: @GreatSoshiant our [pattern guide is here](https://www.notion.so/arbitrum/Pattern-guide-Docs-dafaf8d1ce34433d88c1ef3b4b99c19c?pvs=4#2309f7f5db284126929065ac3c56fef8) for reference; procedural instructions are a perfect candidate for the how-to content type. Some additional rationale / related TODOs are in [this PR](https://github.com/OffchainLabs/arbitrum-docs/pull/799#discussion_r1431788299). Need a hand refactoring these, or would you like to take a stab at it?

**Comment by Alirezaabbasi-hub** [comment]
I am experiencing and learning, this is not for presentation



در چهارشنبه 31 ژانویه 2024 در 10:15 بعدازظهر symbolpunk <
***@***.***> نوشت:

File: arbitrum-docs/launch-orbit-chain/concepts/arbos-upgrade.md
Comment (GreatSoshiant) [reply arbitrum-docs/launch-orbit-chain/concepts/arbos-upgrade.md:2]: That completely makes sense to me! @symbolpunk I was just drafting it to have the content in hand and reviewed by our Engs. I will follow procedural instructions pattern as you mentioned and will get back soon!
Thanks

File: arbitrum-docs/launch-orbit-chain/concepts/arbos-upgrade.md
Comment (GreatSoshiant) [reply arbitrum-docs/launch-orbit-chain/concepts/arbos-upgrade.md:33]: So you mean changing the orders and put update arb nodes as first step @hkalodner ?

**Review by Alirezaabbasi-hub — COMMENTED** [review]
> +The process of upgrading ArbOS on an Orbit chain involves specific steps, which are outlined in the following section.
+
+Before delving into the steps for upgrading, it's important to understand the concept of chain ownership and the methods through which a chain owner can upgrade their chain. These topics are comprehensively covered in the [Chain Ownership](chain-ownership.md) section of the documentation.
+
+To upgrade ArbOS on an Orbit chain, follow these steps:
+
+1. **Wasm Module Root Upgrade**: The `WASM module root` is a 32-byte hash, created from the Merkelized Go replay binary and its dependencies. When ArbOS is upgraded, a new Wasm module root is generated due to modifications in the State Transition Function. This new wasm module root must be set in the rollup contract on the parent chain. To implement the upgrade, you should utilize the [setWasmModuleRoot](https://github.com/OffchainLabs/nitro-contracts/blob/38a70a5e14f8b52478eb5db08e7551a82ced14fe/src/rollup/RollupAdminLogic.sol#L321) method by calling the `Rollup proxy` contract. As the `upgrade executor` contract on the parent chain is the designated owner of the rollup contract, it is necessary for the **chain owner account** to initiate a call to the upgrade executor. This call should include the correct calldata for setting the new wasm module root through the setWasmModuleRoot function within the rollup contract. For example, the wasm module root for ArbOS 11 is `0xf4389b835497a910d7ba3ebfb77aa93da985634f3c052de1290360635be40c4a`. 
+**Note** that wasm module roots are backward compatible, meaning upgrading it prior to an ArbOS upgrade won't disrupt the system.
+
+1. **Schedule ArbOS Upgrade**: To update the ArbOS version, you must schedule an upgrade on the Orbit chain. This is done by calling the [scheduleArbOSUpgrade](https://github.com/OffchainLabs/nitro-contracts/blob/acb0ef919cce9f41da531f8dab1b0b31d9860dcb/src/precompiles/ArbOwner.sol#L61) function on the ArbOwner precompile of the Orbit chain. This function requires two inputs:
+    1. **newVersion**: Specify the ArbOS version you wish to upgrade to.
+    2. **timestamp**: Set the exact timestamp at which you want your Orbit chain to transition to the new ArbOS version.
+
+    As with the parent chain, performing **admin actions** on the orbit chain necessitates using the `upgrade executor` contract on the orbit chain. The **chain owner account** must execute a call to the upgrade executor with the appropriate calldata, to invoke the `scheduleArbOSUpgrade` function of the ArbOwner precompile, thereby setting the new ArbOS version and the desired timestamp for the upgrade. 
+
+3. **Update Arb Nodes**: The final step in upgrading ArbOS is to update the nodes of your Orbit chain to a version that supports the new ArbOS. This update should be performed after completing steps **1** and **2**, and prior to the **timestamp** set for the ArbOS upgrade. For example, if you are upgrading to ArbOS 11, the corresponding node version is `v2.2.2`. The Docker image for this node version is `offchainlabs/nitro-node:v2.2.2-8f33fea`. This step ensures that your nodes are compatible with the new ArbOS version and are ready to operate effectively after the upgrade.

**Comment by symbolpunk** [comment]
In review.

File: arbitrum-docs/launch-orbit-chain/how-tos/arbos-upgrade.md
Comment (symbolpunk) [inline arbitrum-docs/launch-orbit-chain/how-tos/arbos-upgrade.md:42]:
```suggestion
To schedule an ArbOS version upgrade, call the `[scheduleArbOSUpgrade](https://github.com/OffchainLabs/nitro-contracts/blob/acb0ef919cce9f41da531f8dab1b0b31d9860dcb/src/precompiles/ArbOwner.sol#L61)` function on the `ArbOwner` precompile of the Orbit chain(s) you're upgrading. This function requires two inputs:
```

File: arbitrum-docs/launch-orbit-chain/how-tos/arbos-upgrade.md
Comment (symbolpunk) [inline arbitrum-docs/launch-orbit-chain/how-tos/arbos-upgrade.md:43]:
```suggestion
1. **`newVersion`**: Specify the ArbOS version you wish to upgrade to.
```

Can we provide an example of this call, so the reader can be confident about the format of the input params?

File: arbitrum-docs/launch-orbit-chain/how-tos/arbos-upgrade.md
Comment (symbolpunk) [inline arbitrum-docs/launch-orbit-chain/how-tos/arbos-upgrade.md:44]:
```suggestion
2. **`timestamp`**: Set the exact timestamp at which you want your Orbit chain to transition to the new ArbOS version.
```

What format is this timestamp? Is it UTC? Does it matter?

File: arbitrum-docs/launch-orbit-chain/how-tos/arbos-upgrade.md
Comment (symbolpunk) [inline arbitrum-docs/launch-orbit-chain/how-tos/arbos-upgrade.md:46]:
```suggestion
Because this is an administrative action (similar to upgrading your Wasm module root), the **chain owner account** must call the target chain's `upgrade executor` contract with the appropriate calldata in order to invoke the `scheduleArbOSUpgrade` function of the ArbOwner precompile. This will schedule the ArbOS upgrade using the specified version and timestamp. 
```

File: arbitrum-docs/launch-orbit-chain/how-tos/arbos-upgrade.md
Comment (symbolpunk) [inline arbitrum-docs/launch-orbit-chain/how-tos/arbos-upgrade.md:47]: :::tip Immediate upgrades
To upgrade immediately (without scheduling), set the `timestamp` to `0`.
:::

Do we need to specify whether this is a string or int?

File: arbitrum-docs/launch-orbit-chain/how-tos/arbos-upgrade.md
Comment (symbolpunk) [inline arbitrum-docs/launch-orbit-chain/how-tos/arbos-upgrade.md:49]:
```suggestion
Congratulations! You've upgraded your Orbit chain(s) to the specified ArbOS version.
```

One possible way to improve this is to set expectations in the "success" case - something along the lines of "You should see <insert console log or some way to verify that the upgrade has been scheduled successfully>". This way, readers know what to look for if they're not confident that their contract calls were successfully executed.

Not needed, just something to consider as a general pattern (I'm not even sure if this is possible or idiomatic with these contract calls)

File: arbitrum-docs/launch-orbit-chain/how-tos/arbos-upgrade.md
Comment (symbolpunk) [inline arbitrum-docs/launch-orbit-chain/how-tos/arbos-upgrade.md:47]:
```suggestion
:::tip Immediate upgrades
To upgrade immediately (without scheduling), set the timestamp to 0.
:::
```

Do we need to specify whether this is a string or int?

File: arbitrum-docs/launch-orbit-chain/how-tos/arbos-upgrade.md
Comment (GreatSoshiant) [reply arbitrum-docs/launch-orbit-chain/how-tos/arbos-upgrade.md:13]: Completely makes sense to me

File: arbitrum-docs/launch-orbit-chain/how-tos/arbos-upgrade.md
Comment (GreatSoshiant) [reply arbitrum-docs/launch-orbit-chain/how-tos/arbos-upgrade.md:41]: We discussed it with @hkalodner. Because we can't give a general example for this, for this version of doc we would not add an example. We're working on Orbit SDK support and after it's merged we can explain how to use it to generate the calldata

File: arbitrum-docs/launch-orbit-chain/how-tos/arbos-upgrade.md
Comment (GreatSoshiant) [reply arbitrum-docs/launch-orbit-chain/how-tos/arbos-upgrade.md:47]: That's a great suggestion for these type of docs but as you mentioned, for this specific doc, we don't have a success sign, cause it's a complete network upgrade. We can think about it when adding orbit sdk support.
Thanks!

File: arbitrum-docs/launch-orbit-chain/how-tos/arbos-upgrade.md
Comment (GreatSoshiant) [reply arbitrum-docs/launch-orbit-chain/how-tos/arbos-upgrade.md:45]: I don't think so because we said that it's a unix timestamp above

**Review by symbolpunk — CHANGES_REQUESTED** [review]
Additional comments [in the diff tab here](https://github.com/OffchainLabs/arbitrum-docs/pull/919/files) 🙏

File: arbitrum-docs/launch-orbit-chain/how-tos/arbos-upgrade.md
Comment (GreatSoshiant) [reply arbitrum-docs/launch-orbit-chain/how-tos/arbos-upgrade.md:25]: I just added this for now for people that wanna upgrade ArbOS 11 to be sure they know which node version they need to use Mick. I'll delete it after a few weeks so it's good for now

File: arbitrum-docs/launch-orbit-chain/how-tos/arbos-upgrade.md
Comment (GreatSoshiant) [reply arbitrum-docs/launch-orbit-chain/how-tos/arbos-upgrade.md:29]: I refer them to the chain ownership page already. They can find some info here. Also I talked with @hkalodner about it. I think it's better to go with this version for now and we will add more info after Orbit SDK support. For now we can't give more explanation on it that helps generally and also I think for people who we want to share this with, the explanation would be enough for now. But completely agree that we need to add more info after Orbit SDK support

File: arbitrum-docs/launch-orbit-chain/how-tos/arbos-upgrade.md
Comment (GreatSoshiant) [reply arbitrum-docs/launch-orbit-chain/how-tos/arbos-upgrade.md:21]: Actually the target audinece for this doc wouldn't be users Mick. The target audience is RaaS providers so we don't need to specify all details about the steps. I know your concern about giving as more detail as possible but for this doc I think it's enough just to say upgrade your nodes.

File: arbitrum-docs/launch-orbit-chain/how-tos/arbos-upgrade.md
Comment (GreatSoshiant) [reply arbitrum-docs/launch-orbit-chain/how-tos/arbos-upgrade.md:23]: Schedule upgrades is explained in detail on next sections. Also orbit chain node's would be a better term for it I think than Nitro. For instance on node runner section we use "how to run orbit chain node". This is the same node as we used there

File: arbitrum-docs/launch-orbit-chain/how-tos/arbos-upgrade.md
Comment (symbolpunk) [inline arbitrum-docs/launch-orbit-chain/how-tos/arbos-upgrade.md:38]:
```suggestion
To schedule an ArbOS version upgrade, call the [`scheduleArbOSUpgrade`](https://github.com/OffchainLabs/nitro-contracts/blob/acb0ef919cce9f41da531f8dab1b0b31d9860dcb/src/precompiles/ArbOwner.sol#L61) function on the `ArbOwner` precompile of the Orbit chain(s) you're upgrading. This function requires two inputs:
```

Formatting fix

File: arbitrum-docs/launch-orbit-chain/how-tos/arbos-upgrade.md
Comment (symbolpunk) [inline arbitrum-docs/launch-orbit-chain/how-tos/arbos-upgrade.md:3]:
```suggestion
sidebar_label: 'Upgrade ArbOS'
```
nit

**Review by symbolpunk — APPROVED** [review]
LGTM, great work @GreatSoshiant. Approving to unblock, with [one final editorial nit](https://github.com/OffchainLabs/arbitrum-docs/pull/919/files#r1480054387). We can either specify Nitro explicitly there, or just change `orbit` -> `Orbit` if this suggested edit doesn't add value.

**Pull request merged**
